proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,551 @@
+way of passing down actor isolation because it still requires dynamic
+dispatch in order to extract the executor.  It would be better for the
+implementation if we could pass down the exact `UnownedSerialExecutor?`
+value that's needed at runtime.  While we do not currently want to
+encourage programmers to work with values of this type directly because
+of its tricky lifetime semantics, the compiler can manage it fairly
+easily.
+
+All of these downsides could be addressed by adding an attribute
+which causes an entity (including an accessor) to inherit its caller's
+isolation. This would be equivalent to receiving an `isolated` parameter
+with the same value as would be produced by `#isolated`, but it's easier
+to write, can be used in a few places that can't add arbitrary parameters, 
+and may be more efficiently implementable.
+
+### Isolated function types
+
+This proposal is focused on propagating isolation information *into*
+functions, but it's also interesting to look at propagating isolation
+*out* of functions.  Currently, the Swift type system only allows
+function isolation to be expressed in limited ways: functions can be
+declared as isolated to a global actor (e.g. `@MainActor () -> ()`), but
+all other kinds of isolation must be "type-erased", leaving a value
+whose type appears to be non-isolated.  This forces some of the same
+awkward `Sendable` restrictions that this proposal discusses with the
+`sequentialMap` example in the Motivation section.
+
+One way to solve this would be to introduce value-dependent isolated
+function types.  With such a feature, you could declare a value to have
+type, say, `@isolated(myActor) () -> ()`, where `myActor` is a `let`
+constant in the local scope.  This kind of value dependence, however,
+is a large step in complexity for a type system, and it's not a likely
+path for Swift in the foreseeable future.
+
+A more promising approach would be to allow the isolation to be
+statically erased but still make it dynamically recoverable by carrying
+it along in the function value, essentially as an extra value of type
+`(any Actor)?`.  This would look something like `@isolated () -> ()`,
+and it could be used to e.g. dynamically propagate the isolation of
+a function into something like the `Task` initializer so that the task
+can immediately start on the right executor.  This would compose well
+with the features in this proposal because it would be natural to allow
+such functions to be used as `isolated` parameters.  This would be very
+nice for functions like `sequentialMap` that should probably be isolated
+not to their *caller* but to the *function they've been passed*.
+
+## Alternatives Considered
+
+### Allowing isolation to `SerialExecutor` types
+
+This proposal observes that it is more efficient to pass down an
+`UnownedSerialExecutor` value instead of an actor reference and suggests
+that `@inheritsIsolation` be implemented this way.  However, if a function
```

Comment:
Now that `@inheritsIsolation` is moved to "Future directions" without a strawman syntax, this is now the first reference to that attribute; probably just strike this second half of the sentence? Some of the following text may also need to be reworked (for example, there is nothing _but_ the explicit isolation-polymorphism pattern in this proposal now, so the part of the following sentence that says "_if_ a function wants to use" this pattern is inoperative).